Added Windows store file system watcher

karrtikr opened the pull request.

int19h commented on src/client/pythonEnvironments/common/pythonBinariesWatcher.ts:
```diff
@@ -3,19 +3,22 @@
     callback: (type: FileChangeType, absPath: string) => void,
+    executablePattern: string = executable,
 ): void {
+    const patterns = [executablePattern, `*/${executablePattern}`, `*/${binName}/${executablePattern}`];
```
Since this is a regex and not a glob, it needs to be `.*/` - as it is, the star is applied to `^`, which effectively negates it. Although `.*/` will match any number of subdirectories; so if the intent is to match just one, then I think you want `[^/]+/`.

int19h commented on src/client/pythonEnvironments/common/pythonBinariesWatcher.ts:
```diff
@@ -3,19 +3,22 @@
 export function watchLocationForPythonBinaries(
     baseDir: string,
     callback: (type: FileChangeType, absPath: string) => void,
+    executablePattern: string = executable,
```
Is this supposed to be a glob pattern, or a regex? If it's a glob, then it can't just be substituted into a regex and work the same. If it's a regex, then `patterns` are wrong.

karrtikr replied:
Will correct it.
But originally it was intended as a glob which can also be substituted into a regex 😅 if that makes sense. See [this](https://github.com/microsoft/vscode-python/pull/14577/files#r515234602) comment. I figured python binaries basename patterns are simple enough so the glob can be substituted into a regex, for eg. `python3\.[0-9]\.exe`

But it doesn't have to be that way. Will change to use a library to match glob patterns instead of converting to regex. Any recommendations? I'm thinking to use https://www.npmjs.com/package/minimatch or https://www.npmjs.com/package/picomatch

karrtikr commented on src/client/pythonEnvironments/discovery/locators/services/windowsStoreLocator.ts:
```diff
@@ -87,26 +85,28 @@ export async function isWindowsStoreEnvironment(interpreterPath: string): Promis
+ * python2.7.exe
+ * python3.exe
+ * python38.exe
+ * Note the below is constructed in a way that it can be used as a glob pattern, or as a regex.
```
...

int19h commented on src/client/pythonEnvironments/common/pythonBinariesWatcher.ts:
```diff
@@ -3,19 +3,23 @@
     for (const pattern of patterns) {
         watchLocationForPattern(baseDir, pattern, (type: FileChangeType, e: string) => {
-            if (!e.endsWith(executable)) {
+            const isMatch = picomatch(executablePattern);
```
It should also use `nocase: true` on Windows.

karrtikr replied:
For some reason `@types/picomatch` doesn't have `nocase` property defined in [PicomatchOptions](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f49c5606b7b4283042d2bae2e6425f7bbed028bd/types/picomatch/index.d.ts#L10-L19)
But it was in `toRegex` method, so I used that.

karrtikr replied:
The `nocase` property is buggy, for some reason it's not working, removed it.

int19h replied:
But if it's not case-insensitive, it won't always work correctly on Windows. I've seen "Python.exe" and such before.

ericsnowcurrently commented on src/client/pythonEnvironments/common/pythonBinariesWatcher.ts:
```diff
@@ -3,19 +3,23 @@
     for (const pattern of patterns) {
         watchLocationForPattern(baseDir, pattern, (type: FileChangeType, e: string) => {
-            if (!e.endsWith(executable)) {
+            const regex = picomatch.toRegex(executablePattern, { nocase: getOSType() === OSType.Windows });
+            if (!regex.test(path.basename(e))) {
```
Using `path.basename()` implies certain restrictions on `executablePattern` that you are not enforcing.  I recommend either failing if `executablePattern.includes(path.separator)` or allowing the broader globs:

```suggestion
            const regex = picomatch.toRegex(path.join('**', executablePattern), { nocase: getOSType() === OSType.Windows });
            if (!regex.test(e)) {
```

ericsnowcurrently commented on src/client/pythonEnvironments/discovery/locators/services/windowsStoreLocator.ts:
```diff
@@ -87,26 +86,27 @@ export async function isWindowsStoreEnvironment(interpreterPath: string): Promis
 }
 
+/**
+ * This is a glob pattern which matches following file names:
```
The import point is that it matches Windows Store Python.  So:

```suggestion
 * This is a glob pattern which matches Windows Store Python executables:
```

karrtikr replied:
I have renamed the variable name to `windowsStorePythonExes` instead. The point of this comment was to give examples for file names which are matched, so left the doc as it is.

ericsnowcurrently commented on src/client/pythonEnvironments/discovery/locators/services/windowsStoreLocator.ts:
```diff
@@ -71,9 +72,7 @@ export function isForbiddenStorePath(interpreterPath:string):boolean {
-    const localAppDataStorePath = path
-        .normalize(getWindowsStoreAppsRoot())
-        .toUpperCase();
+    const localAppDataStorePath = path.normalize(getWindowsStoreAppsRoot()).toUpperCase();
```
Please leave one operation per line. Otherwise it is easy to miss the additional operations.

ericsnowcurrently commented on src/client/pythonEnvironments/discovery/locators/services/windowsStoreLocator.ts:
```diff
@@ -130,18 +130,21 @@ export async function getWindowsStorePythonExes(): Promise<string[]> {
-    return files
-        .map((filename:string) => path.join(windowsAppsRoot, filename))
-        .filter(isWindowsStorePythonExe);
+    return files.map((filename: string) => path.join(windowsAppsRoot, filename)).filter(isWindowsStorePythonExe);
```
Please leave one operation per line. Otherwise it is harder to read and easy to miss the additional operations.

karrtikr replied:
The formatter did this automatically, will reverse.

ericsnowcurrently commented on src/client/pythonEnvironments/discovery/locators/services/windowsStoreLocator.ts:
```diff
@@ -87,26 +86,27 @@ export async function isWindowsStoreEnvironment(interpreterPath: string): Promis
-
-    return windowsPythonExes.test(path.basename(interpreterPath));
+export function isWindowsStorePythonExe(interpreterPath: string): boolean {
+    const regex = picomatch.toRegex(windowsPythonExes, { nocase: true });
+    return regex.test(path.basename(interpreterPath));
```
Why use globs here instead of regexes?

karrtikr replied:
`windowsPythonExes` is also passed into watcher API which expects a glob for reasons I mentioned [here](https://github.com/microsoft/vscode-python/pull/14577#discussion_r516237474).
I did not want to define `windowsPythonExes` twice once as a regex and other as a glob, hence just a glob.

ericsnowcurrently commented on src/client/pythonEnvironments/common/pythonBinariesWatcher.ts:
```diff
@@ -3,19 +3,23 @@
     for (const pattern of patterns) {
         watchLocationForPattern(baseDir, pattern, (type: FileChangeType, e: string) => {
-            if (!e.endsWith(executable)) {
+            const regex = picomatch.toRegex(executablePattern, { nocase: getOSType() === OSType.Windows });
+            if (!regex.test(path.basename(e))) {
```
Why are we using globs here instead of regexes?  I don't see any advantage.

karrtikr replied:
`executablePattern` has to be passed into `watchLocationForPattern`, which expects a glob pattern, as the underlying VSCode and chokidar APIs use glob patterns. Hence I'm constrained to use globs.

ericsnowcurrently commented on src/client/pythonEnvironments/discovery/locators/services/windowsStoreLocator.ts:
```diff
@@ -130,18 +130,21 @@ export async function getWindowsStorePythonExes(): Promise<string[]> {
+
+    public constructor() {
+        super();
+        this.registerWatchers().ignoreErrors();
```
Constructors should not have side effects (other than setting object state) or start any background operations.  Instead put that stuff in a separate `initialize()` method.

There should also be a `stop()` or `dispose()` method to stop the watcher.

karrtikr replied:
At this point I'm thinking to add an optional `initialize()` in the `ILocator` interface itself. Or I can add it in the [base locator class](https://github.com/microsoft/vscode-python/blob/218dd8de3ac990a3587d2cc157e6f44d18cd9dd7/src/client/pythonEnvironments/base/locator.ts#L183). All the locators will need a watcher anyways, plus we also use `initialize()` method in caching and cache locator. Agree? Perhaps a separate PR?

karrtikr replied:
> There should also be a stop() or dispose() method to stop the watcher.

As of now the stop() method doesn't exist on the general watcher method. Will add a work item so `watchLocationForPythonBinaries` returns a `Disposable`.

ericsnowcurrently commented on src/client/pythonEnvironments/discovery/locators/services/windowsStoreLocator.ts:
```diff
@@ -167,4 +170,11 @@ export class WindowsStoreLocator extends Locator {
         return undefined;
     }
+
+    private async registerWatchers(): Promise<void> {
```
```suggestion
    private async startWatcher(): Promise<void> {
```

Also, why is the method async?

karrtikr replied:
The method returns a promise so that I can start the watcher in background without blocking the locator.
The method is `async` because Typescript complains in case of `private registerWatchers(): Promise<void>`

ericsnowcurrently reviewed (changes requested):
The major comments:

* why are we using globs instead of regexes?
* do not start the watcher in the constructor

karrtikr commented:
@ericsnowcurrently The glob module that we use is built on top of minimatch, so that one is a dependency freebie. We also use that in the extension. Hence directly used that to avoid the disclaimer comments, converting regex to glob and then vice versa without any additional cost.

ericsnowcurrently commented on src/test/pythonEnvironments/discovery/locators/windowsStoreLocator.test.ts:
```diff
@@ -0,0 +1,143 @@
+        let actualEvent: PythonEnvsChangedEvent;
+        const deferred = createDeferred<void>();
+        await windowsStoreEnvs.create();
+        // Wait before the change event has been sent. If both operations occur almost simultaneously no event is sent.
+        await sleep(100);
```
Is there some other way we can wait than to sleep?  Sleeps inevitably lead to flaky tests.

karrtikr replied:
I don't see how. But in this case this is mainly used as a precaution, and 100ms should be good enough.

ericsnowcurrently replied:
> 100ms should be good enough.

Famous last words. 😄

ericsnowcurrently reviewed (changes requested):
Overall this is looking good.  Thanks for the fixes.  I've left a few minor comments in the non-test files.  Otherwise all the rest of the comments are in the test files.

karrtikr commented on src/client/pythonEnvironments/common/pythonBinariesWatcher.ts:
```diff
@@ -3,19 +3,23 @@
     callback: (type: FileChangeType, absPath: string) => void,
+    executableGlob: string = executable,
 ): void {
+    const patterns = [executableGlob, `*/${executableGlob}`, `*/${binName}/${executableGlob}`];
```
I went with your initial suggestion to expand the glob pattern to support more than base patterns. https://github.com/microsoft/vscode-python/pull/14577#discussion_r516197149

So we don't have to put a restriction on the wildcards or path separators. I can rename it to `executableSuffixGlob` to make it clearer.

ericsnowcurrently replied:
> I went with your initial suggestion to expand the glob pattern to support more than base patterns.

I don't see that.  What I see is:

  `const pythonExeGlob = 'python3\.[0-9]*\.exe';`

A more-than-base pattern would be:

  `const pythonExeGlob = '**/python3\.[0-9]*\.exe';`

or:

  `const pythonExeGlob = path.join('**', 'python3\.[0-9]*\.exe');`

karrtikr replied:
I'm talking about the API, not this particular usecase. The API supports more than base patterns, i.e user can pass in a `path.join('**', 'python3\.[0-9]*\.exe')` and things should work.

ericsnowcurrently replied:
`path.join('**', executablesGlob)` is exactly what I would expect this function to do for the caller.

karrtikr replied:
I think you're assuming `executablesGlob` to be the base executable glob, it's not. It refers to the suffix of the full executable path we intend to watch. For eg. it can be  `path.join('**', 'python3\.[0-9]*\.exe')` itself, so the function need not do anything for the caller.

Hence why I suggested to rename it to `executablesSuffixGlob`.

ericsnowcurrently replied:
The patterns on this line suggest otherwise:

* `executableGlob` - look for the file in the current directory
* `*/${executableGlob}` - look for the file one level down
* `*/${binName}/${executableGlob}` - look for the file in the "bin" directory one level down

What you seen to be saying is that a caller may want to search at some arbitrary depth relative to a given root:

* `**/python3.exe`
* `*/**/python3.exe`
* `*/${binName}/**/python3.exe`

However, those don't seem right to me.  Perhaps you meant a user might want to do something like the following?

* `**/python3.exe`
* `**/*/python3.exe`
* `**/*/${binName}/python3.exe`

The problem is that kind of defeats the structure-oriented design of this function.  So I don't see that to be right either.

All this is why I would expect `executableGlob` to always be a base name.

karrtikr replied:
> What you seen to be saying is that a caller may want to search at some arbitrary depth relative to a given root:
> * `**/python3.exe`
> * `*/**/python3.exe`
> * `*/${binName}/**/python3.exe`
> However, those don't seem right to me.

I realize it doesn't seem right. But if you look closely, the first pattern is enough to do what the user asks (watch for suffixes ending with `python3.exe`). And the second and third pattern are just subsets of the first. So yes, they do what the user intended to.

ericsnowcurrently replied:
That is only true if the glob is for the basename.  If they put `**/...` then it won't be what they expect.  If it shouldn't matter then this function should not be auto-watching `*/${executableGlob}` and `*/${binName}/${executableGlob}`.  The two things just don't make sense together.

karrtikr commented on src/client/pythonEnvironments/discovery/locators/services/windowsStoreLocator.ts:
```diff
@@ -131,17 +132,21 @@ export async function getWindowsStorePythonExes(): Promise<string[]> {
+    private readonly kind: PythonEnvKind = PythonEnvKind.WindowsStore;
+
+    public initialize(): void {
+        this.startWatcher();
```
I already have another PR almost ready with the changes. If this is a temporary comment I think we can hold it off until then.

karrtikr commented on src/test/pythonEnvironments/discovery/locators/windowsStoreLocator.test.ts:
```diff
@@ -0,0 +1,143 @@
+import { TEST_TIMEOUT } from '../../../constants';
+import { TEST_LAYOUT_ROOT } from '../../common/commonTestConstants';
+
+const testLocalAppData = path.join(TEST_LAYOUT_ROOT, 'storeApps');
+const testStoreAppRoot = path.join(testLocalAppData, 'Microsoft', 'WindowsApps');
+const testExecutablePath = path.join(testStoreAppRoot, 'python3.4.exe');
```
I was worried about the timeout. Tests where we create environments can take time, and I think functional tests have a timeout of just 3 seconds?

ericsnowcurrently replied:
In "build/.mocha.functional.json": `"timeout": 180000`

karrtikr replied:
Actually the tests rely on VSCode APIs.

karrtikr commented on src/client/pythonEnvironments/discovery/locators/services/windowsStoreLocator.ts:
```diff
@@ -87,26 +88,26 @@ export async function isWindowsStoreEnvironment(interpreterPath: string): Promis
+ * python3.exe
+ * python38.exe
+ */
+const windowsStorePythonExes = 'python3\.[0-9]\.exe';
```
Also, this executable won't work for 3.10 :) Will update

karrtikr commented on src/test/pythonEnvironments/discovery/locators/windowsStoreLocator.test.ts:
```diff
@@ -0,0 +1,143 @@
+        locator = new WindowsStoreLocator();
+        locator.initialize();
+        // Wait for watchers to get ready
+        await sleep(1000);
```
> Is there any other way to know when they are ready than to wait for a full second?

Actually no. The underlying VSCode watcher API doesn't tell us when it's ready.

> Also, I expect we would want the watchers to be ready any time we initialize the locator and not just during tests. So shouldn't locator.initialize() do the waiting?

IMO `initialize()` call shouldn't block on whether the locator watcher is ready or not in order to return envs through `iterEnvs()`. Which is why it's not in the call.

ericsnowcurrently replied:
> Actually no. The underlying VSCode watcher API doesn't tell us when it's ready.

That seems like something they should add to the API.

> IMO initialize() call shouldn't block on whether the locator watcher is ready or not, so that we can return envs through iterEnvs() without waiting on watchers. Which is why it's not in the call.

`initailize()` doesn't have to block.  It can track readiness in the background.  However, you are right that in normal operation we don't need to wait for the watcher.

ericsnowcurrently replied:
That said, using a sleep here is not desirable.  It sounds like we don't have other options though.

karrtikr replied:
> initailize() doesn't have to block. It can track readiness in the background. However, you are right that in normal operation we don't need to wait for the watcher.

Yeah even if it tracks readiness in the background, looks like we still have to wait in the tests. And for practical purposes, we don't really care when the watcher start, except in the tests, so maybe that's why they don't provide the API.

Anyways, I'm considering this as resolved.

karrtikr commented on src/test/pythonEnvironments/discovery/locators/windowsStoreLocator.test.ts:
```diff
@@ -0,0 +1,143 @@
+    teardown(() => windowsStoreEnvs.cleanUp());
+    suiteTeardown(async () => {
+        process.env.LOCALAPPDATA = localAppDataOldValue;
+        await windowsStoreEnvs.cleanUp();
```
I see, will remove.

karrtikr commented on src/test/pythonEnvironments/discovery/locators/windowsStoreLocator.test.ts:
```diff
@@ -0,0 +1,143 @@
+
+    suiteSetup(async () => {
+        process.env.LOCALAPPDATA = testLocalAppData;
+        await windowsStoreEnvs.cleanUp();
```
Yes, it doesn't do any harm to leave it there. Suppose if we cancel debugging the tests midway, it can handle things like that.

But I guess I can add a random suffix.

karrtikr replied:
Yes, it doesn't do any harm to leave it there. Suppose if we cancel debugging the tests midway, it can handle things like that.

But I guess I can add a random suffix instead, but it's just that it looks dirtier to me.

ericsnowcurrently replied:
Let's make sure we don't leave files behind then.

karrtikr replied:
A user can cancel debugging midway, so we cannot ensure that for sure.

karrtikr commented on src/test/pythonEnvironments/discovery/locators/windowsStoreLocator.test.ts:
```diff
@@ -0,0 +1,143 @@
+const testStoreAppRoot = path.join(testLocalAppData, 'Microsoft', 'WindowsApps');
+const testExecutablePath = path.join(testStoreAppRoot, 'python3.4.exe');
+
+class WindowsStoreEnvs {
```
I did have a look, but I don't see it making things clearer for this particular case, so skipping. But definitely something to keep in mind.

karrtikr commented on src/test/pythonEnvironments/discovery/locators/windowsStoreLocator.test.ts:
```diff
@@ -0,0 +1,143 @@
+        await windowsStoreEnvs.cleanUp();
+    });
+    setup(async () => {
+        locator = new WindowsStoreLocator();
+        locator.initialize();
```
I would agree, but in this case there're 3 lines:
```typescript
        locator = new WindowsStoreLocator();
        locator.initialize();
        // Wait for watchers to get ready
        await sleep(1000);
```

To me moving these makes tests more cluttered, and larger. But I can factor this setup out in a `initialize()` method if that would help.

karrtikr replied:
Factored it out in the tests.

karrtikr commented on src/client/pythonEnvironments/common/pythonBinariesWatcher.ts:
```diff
@@ -3,19 +3,23 @@
 export function watchLocationForPythonBinaries(
     baseDir: string,
     callback: (type: FileChangeType, absPath: string) => void,
+    executableGlob: string = executable,
```
https://github.com/microsoft/vscode-python/pull/14577#discussion_r516893654

ericsnowcurrently commented on src/client/pythonEnvironments/discovery/locators/services/windowsStoreLocator.ts:
```diff
@@ -87,26 +88,27 @@ export async function isWindowsStoreEnvironment(interpreterPath: string): Promis
-    const windowsPythonExes = /^python(3(.\d+))\.exe$/;
-
-    return windowsPythonExes.test(path.basename(interpreterPath));
+export function isWindowsStorePythonExe(interpreterPath: string): boolean {
+    return minimatch(path.basename(interpreterPath), pythonExeGlob, { nocase: true });
 }
```
Why aren't we just using `isWindowsPythonExe()`?

karrtikr replied:
Because of the reason mentioned in the sprint demo today: We're matching only `python3.*.exes` so that we do collect only unique executables. `isWindowsPythonExe()` also matches with `python.exe` which is not what we want.

karrtikr replied:
The doc comment explains this.

ericsnowcurrently replied:
Then I'd expect the following to take care of it:

```suggestion
export function isWindowsStorePythonExe(interpreterPath: string): boolean {
    if (!isWindowsPythonExe(interpreterPath)) {
        return false;
    }
    return path.basename(interpreterPath) !== 'python.exe';
}
```

That way we don't accept executables we shouldn't, like "python3-whoa!!!.exe".

karrtikr replied:
Right, there's also `python3.exe`, and `python2.exe` that we would need to ignore. I can either do this or correct the glob pattern to not accept the whoaaa string.

karrtikr replied:
Will change the glob pattern to `python3\.[0-9]*([0-9])\.exe`.

karrtikr replied:
I can use `python3\.[0-9]*([0-9])\.exe`.

ericsnowcurrently replied:
Ah, so we're okay with "extended glob".  What you have looks fine, but I'd change the `*` to `?`.

ericsnowcurrently commented on src/client/pythonEnvironments/discovery/locators/services/windowsStoreLocator.ts:
```diff
@@ -87,26 +88,27 @@ export async function isWindowsStoreEnvironment(interpreterPath: string): Promis
+ * python38.exe
+ * 'python.exe', 'python3.exe', and 'python3.8.exe' can point to the same executable, hence only capture python3.*.exes.
+ */
+const pythonExeGlob = 'python3\.[0-9]*\.exe';
```
Note that this diverges from the regex in `isWindowsPythonExe()`.  If it's only for the sake of `3.10` then perhaps it is evidence of why globs aren't great for this.  You may want to have two globs: one with a single digit and one with two.

karrtikr replied:
Because of the reason mentioned in the sprint demo today: We're matching only `python3.*.exes` so that we do collect only unique executables. `isWindowsPythonExe()` also matches with `python.exe` which is not what we want.

ericsnowcurrently commented on src/client/pythonEnvironments/discovery/locators/services/windowsStoreLocator.ts:
```diff
@@ -167,4 +173,11 @@ export class WindowsStoreLocator extends Locator {
+
+    private startWatcher(): void {
+        const windowsAppsRoot = getWindowsStoreAppsRoot();
+        watchLocationForPythonBinaries(windowsAppsRoot, (type: FileChangeType) => {
+            this.emitter.fire({ type, kind: this.kind });
+        }, pythonExeGlob);
```
```suggestion
        watchLocationForPythonBinaries(
            windowsAppsRoot,
            (type: FileChangeType) => {
                this.emitter.fire({ type, kind: this.kind });
            },
            pythonExeGlob,
        );
```

ericsnowcurrently commented on src/client/pythonEnvironments/discovery/locators/services/windowsStoreLocator.ts:
```diff
@@ -167,4 +173,11 @@ export class WindowsStoreLocator extends Locator {
+        const windowsAppsRoot = getWindowsStoreAppsRoot();
+        watchLocationForPythonBinaries(windowsAppsRoot, (type: FileChangeType) => {
+            this.emitter.fire({ type, kind: this.kind });
+        }, pythonExeGlob);
```
Note that this is really the only place that we need `pythonExeGlob`.  So it may make sense to move it inline here.

karrtikr replied:
We need it in `isWindowsStorePythonExe` as well.

ericsnowcurrently replied:
If we use `isWindowsPythonExe()` there then only need it here.

karrtikr replied:
But then we would have two places where we capture what is a windows store executable? The glob pattern itself and `isWindowsStorePythonExe`.

ericsnowcurrently reviewed (approved):
I only have a few more minor comments.  So...

LGTM once all the open comments have been addressed.

karrtikr commented:
@ericsnowcurrently I believe I have addressed all the comments. I generally like to keep comments open for the reviewers to resolve them themselves. Hope I addressed most of your concerns.
